modules/nlu/src/backend/predict-pipeline.ts

```diff
@@ -80,8 +85,37 @@ async function DetectLanguage(
     detectedLanguage = NA_LANG
   }
 
-  const usedLanguage =
-    detectedLanguage !== NA_LANG && elected.value > threshold ? detectedLanguage : input.defaultLanguage
+  // if ML-based language identifier didn't find a match
+  // we proceed with a custom vocabulary matching algorithm
+  // ie. the % of the sentence comprised of tokens in the training vocabulary
+  if (detectedLanguage === NA_LANG) {
+    try {
+      const match = _.chain(supportedLanguages)
+        .map(lang => ({
+          lang,
+          sentence: input.sentence.toLowerCase(),
+          tokens: _.orderBy(Object.keys(predictorsByLang[lang].vocabVectors), 'length', 'desc')
+        }))
+        .map(({ lang, sentence, tokens }) => {
+          for (const token of tokens) {
+            sentence = sentence.replace(token, '')
+          }
```

Comment:
To make this correct, you should replace consecutive spaces (due to token replacement) and `.trim()` whitespaces

Reply:
spaces are part of the vocab, didn't want to assume spaces were used in that language.  but should def account for multiple times the same token in the sentence tho (but the result will be relatively the same)

Reply:
I just noticed it while trying this out. You're right, results are roughly the same but still lets' say you have the following sentence "Iam feeling lucky and alive" and we have "feeling,lucky,and" in our vocab, you will end up with iam___alive where your score would have been greater with iam_alive.

Anyways, makes no real difference at the end